src/reactjs_jsx_ppx_v3.ml

```diff
@@ -232,6 +233,44 @@ let makeExternalDecl fnName loc namedArgListWithKeyAndRef namedTypeList =
     (makePropsType ~loc namedTypeList)
   [@@raises Invalid_argument]
 
+let rec newtypeToVar newtype typ =
+  let traverse = newtypeToVar newtype in
+  { typ with ptyp_desc =
+    match typ.ptyp_desc with
```

Comment:
How about using `Ast_mapper`. That should take care of most boilerplate.

Reply:
I'm not really sure how to run a mapper on a single type expression, or even a general expression. Looks to me like it only does source to source transformation. But if possible, that would be ideal I think.

Reply:
Looks like this:

```ml
let newtypeToVar newtype type_ =
  let typ (mapper : Ast_mapper.mapper) typ =
    match typ.ptyp_desc with
    | Ptyp_constr ({txt = Lident name}, _) when name = newtype ->
      {typ with ptyp_desc = Ptyp_var newtype}
    | Ptyp_extension _ ->
      Location.raise_errorf
        "extensions are not allowed in @react.component definitions"
    | _ -> Ast_mapper.default_mapper.typ mapper typ
  in
  let mapper = {Ast_mapper.default_mapper with typ} in
  mapper.typ mapper type_
```

Reply:
Another questions is what happens if the original type contains both `a` and `'a`, worth trying to construct such an example to see what happens, that there is no accidental capture.

Reply:
Ah, thanks for the example. Makes much more sense now! Is it still necessary to raise an error with exceptions? I would think the mapper would handle that according to expectations now.

Reply:
Yes it would. Just wanted to keep the same semantics in the example.

Reply:
To illustrate the ast mapper.

Reply:
> Another questions is what happens if the original type contains both a and 'a, worth trying to construct such an example to see what happens, that there is no accidental capture.

Yep, that does happen. Inferred unification variables won't collide, if there's a newtype `a` it will avoid using `'a` for unification variables, but if there's an explicit constraint on `'a` it will collide.

I can think of two ways to address that:
1. Mangle the variable in some way, which makes conflicts much less likely, but not impossible and I worry that it might make errors confusing.
2. Do another pass first to find all type variables in order to pick one that isn't already used.

Reply:
Seems rescript will accept anything these days with `\""`, but definitely makes it more unlikely that there's a conflict at least! And I think using a name like `type-a` will make it pretty obvious what it refers to.